control/grpc.py

```diff
@@ -0,0 +1,445 @@
+#
+#  Copyright (c) 2021 International Business Machines
+#  All rights reserved.
+#
+#  SPDX-License-Identifier: LGPL-3.0-or-later
+#
+#  Authors: anita.shekar@ibm.com, sandy.kaur@ibm.com
+#
+
+import socket
+import grpc
+import json
+import logging
+from google.protobuf import json_format
+from .proto import gateway_pb2 as pb2
+from .proto import gateway_pb2_grpc as pb2_grpc
+
+
+class GatewayService(pb2_grpc.GatewayServicer):
+    """Implements gateway service interface.
+
+    Handles configuration of the SPDK NVMEoF target according to client requests.
+
+    Instance attributes:
+        config: Basic gateway parameters
+        logger: Logger instance to track server events
+        gateway_name: Gateway identifier
+        gateway_state: Methods for target state persistence
+        spdk_rpc: Module methods for SPDK
+        spdk_rpc_client: Client of SPDK RPC server
+    """
+
+    def __init__(self, config, gateway_state, spdk_rpc, spdk_rpc_client):
+
+        self.logger = logging.getLogger(__name__)
+        self.config = config
+        self.gateway_state = gateway_state
+        self.spdk_rpc = spdk_rpc
+        self.spdk_rpc_client = spdk_rpc_client
+
+        self.gateway_name = self.config.get("gateway", "name")
+        if not self.gateway_name:
+            self.gateway_name = socket.gethostname()
+
+    def create_bdev(self, request, context=None):
+        """Creates a bdev from an RBD image."""
+
+        self.logger.info(
+            f"Received request to create bdev {request.bdev_name} from" +
+            f" {request.ceph_pool_name}/{request.rbd_name}" +
+            f" with block size {request.block_size}")
+        try:
+            bdev_name = self.spdk_rpc.bdev.bdev_rbd_create(
+                self.spdk_rpc_client,
+                name=request.bdev_name,
+                pool_name=request.ceph_pool_name,
+                rbd_name=request.rbd_name,
+                block_size=request.block_size,
+            )
+            self.logger.info(f"create_bdev: {bdev_name}")
+        except Exception as ex:
+            self.logger.error(f"create_bdev failed with: \n {ex}")
+            if context:
+                context.set_code(grpc.StatusCode.INTERNAL)
+                context.set_details(f"{ex}")
+            return pb2.req_status()
+
+        if context:
+            # Update gateway state
+            try:
+                json_req = json_format.MessageToJson(
+                    request, preserving_proto_field_name=True)
+                self.gateway_state.add_bdev(bdev_name, json_req)
+            except Exception as ex:
+                self.logger.error(
+                    f"Error persisting create_bdev {bdev_name}: {ex}")
+                raise
+
+        return pb2.req_status(status=True)
```

Comment:
Yes. Is there some case where the returned bdev name would not match?

Reply:
It feels like there shouldn't be but that was my question to you ;)

Reply:
As far as I can tell, it will always match. Returning the bdev name was a holdover from an earlier commit where supplying the name in the cli command was optional and SPDK would generate the name. We switched to requiring it to make it easier for [restore](https://github.com/ceph/ceph-nvmeof/pull/9#discussion_r806796622), although now it seems like we could handle it the way we handle nsid and make it optional for the user again. This could be a separate small PR.